tamr_toolbox/low_latency/llm.py

```diff
@@ -0,0 +1,181 @@
+        f"&useManualClustering={do_use_manual_clustering}"
+    )
+    response = tamr_client.post(url)
+    if not response.ok:
+        message = f"Update LLM Failed at submission time: {response.content}"
+        LOGGER.error(message)
+        raise RuntimeError(message)
+    operation_id = response.content.decode("latin1")
+
+    # An operation id of '-1' is returned when LLM is already up to date
+    if operation_id != "-1":
+        operation = Operation.from_resource_id(tamr_client, operation_id=operation_id)
+        operation.wait()
+    return None
+
+
+def poll_llm_status(match_client: Client, *, project_name: str, num_tries: int = 10) -> bool:
+    """
+    Check if LLM is queryable. Try up to num_tries times at 1s intervals.
+
+    Args:
+        match_client: a Tamr client set to use the port of the Match API
+        project_name: name of target mastering project
+        num_tries: optional, max number of times to poll endpoint
+    Returns:
+        bool indicating whether project is queryable
+    """
+
+    url = f"/api/v1/projects/{project_name}:isQueryable"
+    counter = 0
+
+    # Poll endpoint to until project is queryable or num_tries reached
+    while counter < num_tries:
+
+        response = match_client.get(url)
+        queryable = response.content == b"true"
+
+        if queryable:
+            break
+
+        counter += 1
+        time.sleep(1)  # call api at 1s interval if project isn't yet queryable
+
+    return queryable
+
+
+
+def llm_query(
+    match_client: Client,
+    *,
+    project_name: str,
+    records: List[JsonDict],
+    type: Literal['records', 'clusters'],
+    batch_size: int = 1000, 
+    min_match_prob: Optional[float] = None,
+    max_num_matches: Optional[int] = None,
+) -> Dict[int, List[JsonDict]]:
+    """
+    Find the best matching clusters or records for each supplied record. Returns empty list if 
+    response is null.
+
+    Args:
+        match_client: a Tamr client set to use the port of the Match API
+        project_name: name of target mastering project
+        records: list of records to match
+        type: whether to pull record or cluster matches
+        batch_size: split input into this batch size for LLM calls (e.g. to prevent network
+            timeouts), Default 1000
+        min_match_prob: if set, only matches with probability above minimum will be returned, 
+            Default None
+        max_num_matches: if set, at most max_num_matches will be returned for each input record in 
+            records, Default None
+    Returns:
+        Dict keyed by integers (the indices of the records), with value a list containing closest
+            matched clusters
+    Raises:
+        Runtime error if query fails
+    """
+
+    result_dict = defaultdict(lambda: [])  # dict which defaults to empty list to hold results
+
+    url = f"/api/v1/projects/{project_name}:match?type={type}"
+
```

Comment:
Interesting to know that know JS else if statements were similar to Python's

Reply:
This is actually not an else if in the way it looks.  The : and ? are just part of the url. `f"/api/v1/projects/{project_name}:match?type={type}"` evaluates to a string like "/api/v1/projects/my_project:match?type=my_type".
To understand this a bit more, the path `/api/v1/projects/{project_name}` refers to a project object, the `:match` refers to executing a "match" operation on it, and `?type={type}` sets a parameter of the request.

Reply:
Sorry, meant the below code on line 122

Reply:
Python one-line if-else looks like this:
```python
x = 1 if y == 3 else 2  #  if the variable y is 3, assign the variable x the value 1, otherwise assign it the value 2
```